ADR/adr-0008-environment-provisioning.md

```diff
@@ -0,0 +1,464 @@
+- Second, it represents a request from the user to **provision a new deployment target**. Back in the KCP design, a new [Environment] caused the environment controller(s) to create a new sub-workspace, initialize it, and report back a serviceaccount kubeconfig to be used by Argo to administer it.
+
+Some use cases to consider for [Environments]:
+
+1.  As a part of the StoneSoup workspace initialization process, the user should find that both a **dev and stage environment** with corresponding deployment targets are ready for them ([STONE-180](https://issues.redhat.com/browse/STONE-180)). In our post-KCP architecture, these will be backed by **namespaces** on a devsandbox member cluster.
+2.  The user will want to manually create **additional** [Environments] (for example, a prod environment). The user may want to use our compute resources provided in the form of a new **namespace on a devsandbox member cluster** for this ([STONE-183](https://issues.redhat.com/browse/STONE-183)) or they may want to **bring their own cluster** as a target ([STONE-162](https://issues.redhat.com/browse/STONE-162)).
+3.  The integration-service expects to be able to create **ephemeral** [Environments] for automated testing purposes ([STONE-114](https://issues.redhat.com/browse/STONE-114)). For our short-term goals, the automated testing use case requires the same kind of compute as for the dev and stage [Environments] (devsandbox member cluster namespaces), but will expand to include other kinds of deployment targets in the future - like hypershift clusters ([STONE-185](https://issues.redhat.com/browse/STONE-185)).
+
+## Decision
+
+We are going split the two original purposes of the [Environment] CR into different APIs.
+
+- The [Environment] will still represent the first purpose of requesting that HAS and the GitOps service **recognize a new deployment destination**.
+- For the second purpose of **provisioning a new deployment target** we are going to introduce a new [DeploymentTarget] and [DeploymentTargetClaim] API, which will be referenced from an [Environment].
+
+The new [DeploymentTarget] API is designed to emulate storage management APIs (see [persistent-volumes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/) and [storage provisioner design](https://github.com/kubernetes-csi/external-provisioner/blob/master/doc/design.md) for reference).
+
+### CRDs
+
+#### DeploymentTarget (DT)
+
+A deployment target, usually a K8s api endpoint. The credentials for connecting to the target will
+be stored in a secret which will be referenced in the clusterCredentialsSecret field. A [DT] Can be
+created manually by a user, or dynamically using a provisioner. The `phase` section shows the lifecycle phase of the [DT].
+
+**Immutable object**: no
+
+**Scope**: namespace
+
+```yaml
+apiVersion: appstudio.redhat.com/v1alpha1
+kind: DeploymentTarget
+metadata:
+    name: prod-dt
+spec:
+    deploymentTargetClassName: isolation-level-namespace
```

Comment:
It is because in the general case, the `binder` controller's job is to find matching DTs for DTCs, and bind them together. In theory, a DT could be pre-created but sitting around idle. You should be able to create a DTC that requests a certain kind of DT. If there is an existing DT that matches and if it is not yet bound to any other DTCs, then the binder controller should bind _it_ to the DTC.

In order to do that matching, DTs need to declare what storage class they are created from. I.e., if the DTC wants `isolation-level-namespace`, binding controller needs to look for any unbound DTs that provide `isolation-level-namespace`.

Reply:
Got it, sounds very reasonable to me.